include/cuco/detail/trie/bit_vector/bit_vector.cuh

```diff
@@ -0,0 +1,264 @@
+  using selects_storage_type =
+    aow_storage<size_type, 1, extent<size_type>, allocator_type>;  ///< storage type for selects
+
+  /**
+   *@brief Struct to hold all storage refs needed by bitvector_ref
+   */
+  struct device_storage_ref {
+    using size_type       = size_type;     ///< Size type
+    using slot_type       = slot_type;     ///< Slot type
+    using bit_vector_type = bit_vector<>;  ///< bit_vector_ref needs this to access words_per_block
+
+    typename words_storage_type::ref_type words_ref_;  ///< Words ref
+
+    typename ranks_storage_type::ref_type ranks_ref_;      ///< Ranks refs
+    typename selects_storage_type::ref_type selects_ref_;  ///< Selects refs
+
+    typename ranks_storage_type::ref_type ranks0_ref_;      ///< Ranks refs for 0 bits
+    typename selects_storage_type::ref_type selects0_ref_;  ///< Selects refs 0 bits
+  };
+
+  template <typename... Operators>
+  using ref_type =
+    bit_vector_ref<device_storage_ref, Operators...>;  ///< Non-owning container ref type
+
+  /**
+   * @brief Get device ref with operators.
+   *
+   * @tparam Operators Set of `cuco::op` to be provided by the ref
+   *
+   * @param ops List of operators, e.g., `cuco::bv_read`
+   *
+   * @return Device ref of the current `bit_vector` object
+   */
+  template <typename... Operators>
+  [[nodiscard]] auto ref(Operators... ops) const noexcept;
+
+  /**
+   * @brief Get the number of bits bit_vector holds
+   *
+   * @return Number of bits bit_vector holds
+   */
+  size_type constexpr size() const noexcept { return n_bits_; }
+
+  static constexpr size_type words_per_block = 4;  ///< Tradeoff between space efficiency and perf.
+
+ private:
+  size_type n_bits_;  ///< Number of bits bit_vector currently holds
+
+  // These could be public if needed by other classes. Private for now
+  static constexpr size_type bits_per_word  = sizeof(slot_type) * 8;            ///< Bits in a word
+  static constexpr size_type bits_per_block = words_per_block * bits_per_word;  ///< Trivial
+
+  // Host-side structures
+  std::vector<slot_type> words_;  ///< Words vector that represents all bits
+
+  // Device-side structures
+  thrust::device_vector<slot_type> d_words_;   ///< Device words vector
+  thrust::device_vector<rank> ranks_;          ///< Rank values for every 256-th bit (4-th word)
+  thrust::device_vector<rank> ranks0_;         ///< Same as ranks_ but for `0` bits
+  thrust::device_vector<size_type> selects_;   ///< Block indices of (0, 256, 512...)th `1` bit
+  thrust::device_vector<size_type> selects0_;  ///< Same as selects_, but for `0` bits
+
+  allocator_type allocator_;  ///< Allocator used to (de)allocate temporary storage
+  std::unique_ptr<words_storage_type> aow_words_;  ///< Array of window storage structure
+  std::unique_ptr<ranks_storage_type> aow_ranks_;
+  std::unique_ptr<ranks_storage_type> aow_ranks0_;
+  std::unique_ptr<selects_storage_type> aow_selects_;
+  std::unique_ptr<selects_storage_type> aow_selects0_;
```

Comment:
Maybe a leftover, one member group of those two should be redundant.

Reply:
We could remove the aow_storage versions. Using thrust::device_vector throughout would be much simpler. Bitvector class is not using any window functionality like say static_set would.

Reply:
sure, works for me.